Add new CodeMod CombineIsinstanceIssubclass. isistance(x, str) or isnstance(x, (bytes, list)) -> isinstance(x, (str, bytes, list))

LucasFaudman opened the pull request:
### `CombineIsinstanceIssubclass`
- New CodeMod very similar to `CombineStartswithEndswith` but for `isinstance` and `issubclass` calls.
```python
# Input
isinstance(a, int) or isinstance(a, float) or isinstance(a, str)
issubclass(a, str) or issubclass(a, (bytes, bytearray)) or issubclass(a, memoryview)

# Output
isinstance(a, (int, float, str))
issubclass(a, (str, bytes, bytearray, memoryview))
```

Commit:
Add new CodeMod CombineIsinstanceIssubclass. isistance(x, str) or isinstance(x, (bytes, list)) -> isinstance(x, (str, bytes, list))

clavedeluna commented on src/core_codemods/combine_isinstance_issubclass.py:
```diff
@@ -0,0 +1,99 @@
+        self, original_node: cst.BooleanOperation, updated_node: cst.BooleanOperation
+    ) -> cst.CSTNode:
+        if self.matches_isinstance_issubclass_or_pattern(original_node):
+            if not self.filter_by_path_includes_or_excludes(
```
move this filter check up higher or combine if statements somehow.

LucasFaudman replied:
Sounds good I'll look at the core_codemods/docs and make these changes later today. 

Thanks for checking this out!

clavedeluna commented on src/core_codemods/combine_isinstance_issubclass.py:
```diff
@@ -0,0 +1,99 @@
+
+            self.report_change(original_node)
+
+            elements = []
```
same comment as for the combine startswith codemod, worth separating lines 29-44 into a separate method. Maybe this is also pretty much the same as the other codemod and they can share the same function / method?

clavedeluna commented on src/core_codemods/combine_isinstance_issubclass.py:
```diff
@@ -0,0 +1,99 @@
+    ) -> bool:
+        # Match the pattern: isinstance(x, cls1) or isinstance(x, cls2) or isinstance(x, cls3) or ...
+        # and the same but with issubclass
+        args = [m.Arg(value=m.Name()), m.Arg(value=m.Name() | m.Tuple())]
```
it's possible you can more cleanly do this entire match without libcst matchers and instead with a match / case statement. Check out `infer_expression_type` for a good example of how to write that.

clavedeluna replied:
I'd love this comment to be addressed or at least push back explaining if it's not feasible.

LucasFaudman replied:
@clavedeluna Just noticed that I missed this comment. I see what you are saying about the case statement in `utils.infer_expression_type`. 

Should I rewrite it with a case statement now?

LucasFaudman replied:
Funny I just saw your made a comment about this at the exact same time I did. I  have an interview at 10, but I'll address this right after.

clavedeluna replied:
yes, case statements are a bit nicer than isinstance statements especially with so many. let's see what it looks like

LucasFaudman replied:
I tried re-writing with a `match node: case ...` statement like in `infer_expression_type` and I think that that using `libcst.matchers.matches` here produces much more readable code since I am matching on node `shape` not attr values as described in the `libcst.matchers.matches` docs. 

`infer_expression_type` is matching a node by both shape and attr values which I believe is why a case statement make more sense there. 

To do this in `matches_isinstance_issubclass_or_pattern` would require a case for every cst.Node that matches the `isinstance_or` `issubclass_or` or `chained_or` shape and returning 
``` 
node.left.func.value == node.right.func.value  # Same function 
and node.left.args[0].value.value
== node.right.args[0].value.value  # Same first argument (instance)
``` 
for the `isinstance_or | issubclass_or` case 

and 
```
all(
                (
                    call.func.value == node.right.func.value  # Same function
                    and call.args[0].value.value
                    == node.right.args[0].value.value  # Same first argument (instance)
                )
                for call in extract_boolean_operands(node, ensure_type=cst.Call)
            )
```
for the `chained_or` case. 

As far as I can tell this wouldn't be possible to do like in `infer_expression_type` since the attr values are not known and a cst.Node to match with cannot be defined.

What do you think?

clavedeluna replied:
Alright yeah I see what you mean, thanks for indulging me, we can definitely go with the existing implementation

clavedeluna reviewed (changes requested):
Awesome codemod, thanks for adding it!

You'll need to add a few more files such as documentation files. You'll need a file under core_codemods/docs as well as update generate_docs, add an integration_test which will also lead you to register this codemod in the registry. If you have any questions I can give more specific instructions.

drdavella commented:
@LucasFaudman before we merge any of your PRs we need to ask you to acknowledge the contributor license agreement. Unfortunately the CLA bot isn't currently working on this repo due to a bug. But I'll send you an email with the agreement and you can reply with your acknowledgement, if that's okay.

LucasFaudman commented:
> generate_docs

Sure sounds good. I'll send you a message on LinkedIn with my email.

LucasFaudman commented:
> Awesome codemod, thanks for adding it!
> 
> You'll need to add a few more files such as documentation files. You'll need a file under core_codemods/docs as well as update generate_docs, add an integration_test which will also lead you to register this codemod in the registry. If you have any questions I can give more specific instructions.

I know I need to: 
- add DocsMetadata in ./src/codemodder/scripts/generate_docs.py
- add a test to integration_tests (I will model it after test_combine_startswith_endswith.py since both mods are similar)
- add a .md description to ./src/core_codemods/docs/

Is that all?

clavedeluna commented:
> Is that all?

I believe so, but the best way to ensure is look for one codemod to compre. So if you look for `fix-dataclass-defaults` in code and any file with `fix_dataclass_defaults` you'll see all the mentions and files that codemod has

LucasFaudman committed:
Merge branch 'main' into combine-isinstance-issubclass

Commit:
refactor in same way as clavedeluna requests for combine_startswith_endswith, add docs Metadata & .md file, register codemod in __init__

LucasFaudman commented:
@clavedeluna This PR is now ready for review as well.

clavedeluna reviewed (commented):
let's wait to merge the other PR for startswith/endswith that has the same utility func that way you can rebase off that

LucasFaudman committed:
Merge branch 'pixee:main' into combine-isinstance-issubclass

clavedeluna commented on src/core_codemods/combine_isinstance_issubclass.py:
```diff
@@ -0,0 +1,103 @@
+
+        return False
+
+    def make_new_call_from_boolean_operation(
```
alright so now that the other codemod PR was merged, I think you can dedupe this and move to a utility location.

LucasFaudman replied:
I wanted to do this as well but `make_new_call_from_boolean_operation` in `combine_startswith_endswith.py` and `combine_isinstance_issubclass.py` are actually a bit different which is why I didn't generalize and add to `utils.py`:

The differences are:
- `startswith` considers only strings so the `evaluated_value` instead of `value` should be considered to avoid duplicate args if the same arg is passed as both a `SimpleString` and `ConcatenatedString` for example. 
- `startswith` the tuple or arg to combine is `call.args[0].value` For `isinstance` the tuple or arg to combine is `call.args[1].value`.
- `startswith` the instance is part of the Call object. For `isinstance` the instance is `call.args[0].value` (one of the args preceding the combined arg)

To generalize this I could rewrite with kwargs `args_to_insert`, `args_to_combine` and `dedupe_attr` and pass the corresponding values but I thought this would add extra complexity when accounting for arg order, for little reusability unless other codeMods will require this type of logic, then its probably worth it. 

What do you think you?

```python3
#combine_startswith_endswith.py
    def make_new_call_from_boolean_operation(
        self, updated_node: cst.BooleanOperation
    ) -> cst.Call:
        elements = []
        seen_evaluated_values = set()
        for call in extract_boolean_operands(updated_node, ensure_type=cst.Call):
            arg_value = call.args[0].value
            arg_elements = (
                arg_value.elements
                if isinstance(arg_value, cst.Tuple)
                else (cst.Element(value=arg_value),)
            )

            for element in arg_elements:
                if (
                    evaluated_value := getattr(element.value, "evaluated_value", None)
                ) in seen_evaluated_values:
                    # If an element has a non-None evaluated value that has already been seen, continue to avoid duplicates
                    continue
                if evaluated_value is not None:
                    seen_evaluated_values.add(evaluated_value)
                elements.append(element)

        new_arg = cst.Arg(value=cst.Tuple(elements=elements))
        return cst.Call(func=call.func, args=[new_arg])
```

```python3
#combine_isinstance_issubclass.py
    def make_new_call_from_boolean_operation(
        self, updated_node: cst.BooleanOperation
    ) -> cst.Call:
        elements = []
        seen_values = set()
        for call in extract_boolean_operands(updated_node, ensure_type=cst.Call):
            class_or_tuple_arg_value = call.args[1].value
            arg_elements = (
                class_or_tuple_arg_value.elements
                if isinstance(class_or_tuple_arg_value, cst.Tuple)
                else (cst.Element(value=class_or_tuple_arg_value),)
            )

            for element in arg_elements:
                if (value := getattr(element.value, "value", None)) in seen_values:
                    # If an element has a non-None value that has already been seen, continue to avoid duplicates
                    continue
                if value is not None:
                    seen_values.add(value)
                elements.append(element)

        instance_arg = updated_node.left.args[0]
        new_class_or_tuple_arg = cst.Arg(value=cst.Tuple(elements=elements))
        return cst.Call(func=call.func, args=[instance_arg, new_class_or_tuple_arg])
```

clavedeluna replied:
Yes I think you have the right idea. Here's what I came up with. YOu may need to tweak as I didn't test it

```
def make_new_call_from_boolean_operation(self, updated_node: cst.BooleanOperation, arg_index: int, attribute: str, use_original_arg=False) -> cst.Call:
    elements = []
    seen_values = set()
    for call in extract_boolean_operands(updated_node, ensure_type=cst.Call):
        arg_value = call.args[arg_index].value
        arg_elements = (
            arg_value.elements
            if isinstance(arg_value, cst.Tuple)
            else (cst.Element(value=arg_value),)
        )
        for element in arg_elements:
            unique_value = getattr(element.value, attribute, None)
            if unique_value in seen_values:
                continue
            if unique_value is not None:
                seen_values.add(unique_value)
            elements.append(element)

    new_arg = cst.Arg(value=cst.Tuple(elements=elements))
    new_args = [updated_node.left.args[0], new_arg] if use_original_arg else [new_arg]
    return cst.Call(func=call.func, args=new_args)
```

LucasFaudman replied:
Awesome this will definitely work, thank you. I'll have some time to work on this later today

I also was thinking while testing Pixee last night that both codeMods could be even better by supporting more complex boolean operations. Previous implementations required a series of ors of the same call to be merged, but it could support merging 2 or more of the same call anywhere in a boolean operation chain, regardless of the operands before or after the mergeable calls.

For example:
```python3
# before 
x and x.startswith('foo') or x.startswith('bar')
# after
x and x.startswith(('foo', 'bar'))
```
I have a few ideas for how to implement this while maintaining current behavior and I'll test it with something like this. 
```python3
    @each_func
    @pytest.mark.parametrize(
        "input_code, expected",
        [
            # same name and/or on left            
            ("x and x.{func}('foo') or x.{func}('bar')", "x and x.{func}(('foo', 'bar'))"),
            ("x or x.{func}('foo') or x.{func}('bar')", "x or x.{func}(('foo', 'bar'))"),

            # same name and/or on right
            ("x.{func}('foo') or x.{func}('bar') and x", "x.{func}(('foo', 'bar')) and x"),
            ("x.{func}('foo') or x.{func}('bar') or x", "x.{func}(('foo', 'bar')) or x"),
            
            # other name and/or on left
            ("y and x.{func}('foo') or x.{func}('bar')", "y and x.{func}(('foo', 'bar'))"),
            ("y or x.{func}('foo') or x.{func}('bar')", "y or x.{func}(('foo', 'bar'))"),
            
            # other name and/or on right
            ("x.{func}('foo') or x.{func}('bar') and y", "x.{func}(('foo', 'bar')) and y"),
            ("x.{func}('foo') or x.{func}('bar') or y", "x.{func}(('foo', 'bar')) or y"),

            # same name and/or on left, other name and/or on right
            ("x or x.{func}('foo') or x.{func}('bar') or y", "x or x.{func}(('foo', 'bar')) or y"),
            ("x and x.{func}('foo') or x.{func}('bar') or y", "x and x.{func}(('foo', 'bar')) or y"),

            # other name and/or on left, same name and/or on right
            ("y or x.{func}('foo') or x.{func}('bar') or x", "y or x.{func}(('foo', 'bar')) or x"),
            ("y and x.{func}('foo') or x.{func}('bar') or x", "y and x.{func}(('foo', 'bar')) or x"),
        ],
    )
    def test_mixed_boolean_operation(self, tmpdir, func, input_code, expected):
        self._format_func_run_test(tmpdir, func, input_code, expected)    
```

clavedeluna replied:
Awesome! You can add it to this PR or create an issue with all this info to be worked on later

LucasFaudman replied:
I implemented a cleaner solution last night that doesn't require using the recursive func to extract operands and handles all cases regardless of surrounding boolean operands (all previous tests, test above, some new no_change tests). This is the version for startwiths, I'll update the PRs once I generalize and dedupe shared logic from both mods.

`combine_calls` handles the logic that was formerly in `make_new_call_from_boolean_operation` so I will generalize with the same idea you had in the previous message.

Edit; rename for clarity
```python3
class CombineStartswithEndswith(SimpleCodemod, NameResolutionMixin):
    metadata = Metadata(
        name="combine-startswith-endswith",
        summary="Simplify Boolean Expressions Using `startswith` and `endswith`",
        review_guidance=ReviewGuidance.MERGE_WITHOUT_REVIEW,
        references=[],
    )
    change_description = "Use tuple of matches instead of boolean expression"

    
    def leave_BooleanOperation(
        self, original_node: cst.BooleanOperation, updated_node: cst.BooleanOperation
    ) -> cst.CSTNode:
        if not self.filter_by_path_includes_or_excludes(
            self.node_position(original_node)
        ):
            return updated_node

        for call_matcher in map(self.make_call_matcher, ("startswith", "endswith")):
            if self.matches_call_or_call(updated_node, call_matcher):
                self.report_change(original_node)
                return self.combine_calls(updated_node.left, updated_node.right)

            if self.matches_call_or_boolop(updated_node, call_matcher):
                self.report_change(original_node)
                return self.combine_call_or_boolop_fold_right(updated_node)

            if self.matches_boolop_or_call(updated_node, call_matcher):
                self.report_change(original_node)
                return self.combine_boolop_or_call_fold_left(updated_node)

        return updated_node


    def make_call_matcher(self, func_name: str) -> m.Call:
        args = [
            m.Arg(
                value=m.Tuple()
                | m.SimpleString()
                | m.ConcatenatedString()
                | m.FormattedString()
                | m.Name()
            )
        ]

        return m.Call(
            func=m.Attribute(value=m.Name(), attr=m.Name(func_name)),
            args=args,
        )


    def matches_call_or_call(
        self, node: cst.BooleanOperation, call_matcher: m.Call
    ) -> bool:
        # Check for simple case: x.startswith("...") or x.startswith("...")
        call_or_call = m.BooleanOperation(
            left=call_matcher, operator=m.Or(), right=call_matcher
        )
        return (
            m.matches(node, call_or_call)  # Same Func
            and node.left.func.value.value == node.right.func.value.value
        )  # Same Instance

    def matches_call_or_boolop(
        self, node: cst.BooleanOperation, call_matcher: m.Call
    ) -> bool:

        # Check for case when call on left and call on left of right boolop can be combined like:
        # x.startswith("...") or x.startswith("...") and/or <any>
        call_or_boolop = m.BooleanOperation(
            left=call_matcher,
            operator=m.Or(),
            right=m.BooleanOperation(left=call_matcher),
        )
        return (
            m.matches(node, call_or_boolop)  # Same Func
            and node.left.func.value.value == node.right.left.func.value.value
        )  # Same Instance

    def matches_boolop_or_call(
        self, node: cst.BooleanOperation, call_matcher: m.Call
    ) -> bool:

        # Check for case when call on right and call on right of left boolop can be combined like:
        # <any> and/or x.startswith("...") or x.startswith("...")
        boolop_or_call = m.BooleanOperation(
            left=m.BooleanOperation(right=call_matcher),
            operator=m.Or(),
            right=call_matcher,
        )
        return (
            m.matches(node, boolop_or_call)  # Same Func
            and node.left.right.func.value.value == node.right.func.value.value
        )  # Same Instance

    def combine_calls(self, *calls: cst.Call) -> cst.Call:
        elements = []
        seen_evaluated_values = set()
        for call in calls:
            arg_value = call.args[0].value
            arg_elements = (
                arg_value.elements
                if isinstance(arg_value, cst.Tuple)
                else (cst.Element(value=arg_value),)
            )

            for element in arg_elements:
                if (
                    evaluated_value := getattr(element.value, "evaluated_value", None)
                ) in seen_evaluated_values:
                    # If an element has a non-None evaluated value that has already been seen, continue to avoid duplicates
                    continue
                if evaluated_value is not None:
                    seen_evaluated_values.add(evaluated_value)
                elements.append(element)

        new_arg = cst.Arg(value=cst.Tuple(elements=elements))
        return cst.Call(func=call.func, args=[new_arg])

    def combine_call_or_boolop_fold_right(
        self, node: cst.BooleanOperation
    ) -> cst.BooleanOperation:
        new_left = self.combine_calls(node.left, node.right.left)
        new_right = node.right.right
        return cst.BooleanOperation(
            left=new_left, operator=node.right.operator, right=new_right
        )

    def combine_boolop_or_call_fold_left(
        self, node: cst.BooleanOperation
    ) -> cst.BooleanOperation:
        new_left = node.left.left
        new_right = self.combine_calls(node.left.right, node.right)
        return cst.BooleanOperation(
            left=new_left, operator=node.left.operator, right=new_right
        )
```

Commit:
abstract logic from CombineStartswithEndswith and CombineIsinstanceIssubclass into CombineCallsBaseCodemod and make each a subclass, remove unused extract_boolean_operands from utils.py, add more params to test_no_change and test_mixed_boolean_operation for each

LucasFaudman commented:
- `isinstance`, `issubclass`, `startswith` and `endswith` all now will be combined regardless of surrounding operators in a boolean operation chain. 

- All boolean operation and call merging logic previously in `CombineIsinstanceIssubclass` and `CombineStartswithEndswiths` now abstracted into `CombineCallsBaseCodemod` to deduplicate as @clavedeluna asked

- `extract_boolean_operands` no longer needed and removed from utils.py, 

- add more params to `test_no_change` and added `test_mixed_boolean_operation` test for each mod

```python3
class CombineIsinstanceIssubclass(CombineCallsBaseCodemod):
    metadata = Metadata(
        name="combine-isinstance-issubclass",
        summary="Simplify Boolean Expressions Using `isinstance` and `issubclass`",
        review_guidance=ReviewGuidance.MERGE_WITHOUT_REVIEW,
        references=[],
    )
    change_description = "Use tuple of matches instead of boolean expression with `isinstance` or `issubclass`"

    combinable_funcs = ["isinstance", "issubclass"]
    dedupilcation_attr = "value"
    args_to_combine = [1]
    args_to_keep_as_is = [0]

    def make_call_matcher(self, func_name: str) -> m.Call:
        return m.Call(
            func=m.Name(func_name),
            args=[m.Arg(value=m.Name()), m.Arg(value=m.Name() | m.Tuple())],
        )

    def check_calls_same_instance(
        self, left_call: cst.Call, right_call: cst.Call
    ) -> bool:
        return left_call.args[0].value.value == right_call.args[0].value.value


class CombineStartswithEndswith(CombineCallsBaseCodemod):
    metadata = Metadata(
        name="combine-startswith-endswith",
        summary="Simplify Boolean Expressions Using `startswith` and `endswith`",
        review_guidance=ReviewGuidance.MERGE_WITHOUT_REVIEW,
        references=[],
    )
    change_description = "Use tuple of matches instead of boolean expression"

    combinable_funcs = ["startswith", "endswith"]
    dedupilcation_attr = "evaluated_value"
    args_to_combine = [0]
    args_to_keep_as_is = []

    def make_call_matcher(self, func_name: str) -> m.Call:
        return m.Call(
            func=m.Attribute(value=m.Name(), attr=m.Name(func_name)),
            args=[
                m.Arg(
                    value=m.Tuple()
                    | m.SimpleString()
                    | m.ConcatenatedString()
                    | m.FormattedString()
                    | m.Name()
                )
            ],
        )

    def check_calls_same_instance(
        self, left_call: cst.Call, right_call: cst.Call
    ) -> bool:
        return left_call.func.value.value == right_call.func.value.value


class CombineCallsBaseCodemod(SimpleCodemod, NameResolutionMixin):
    combinable_funcs: list[str] = []
    dedupilcation_attr: str = "value"
    args_to_combine: list[int] = [0]
    args_to_keep_as_is: list[int] = []

    def leave_BooleanOperation(
        self, original_node: cst.BooleanOperation, updated_node: cst.BooleanOperation
    ) -> cst.CSTNode:
        if not self.filter_by_path_includes_or_excludes(
            self.node_position(original_node)
        ):
            return updated_node

        for call_matcher in map(self.make_call_matcher, self.combinable_funcs):
            if self.matches_call_or_call(updated_node, call_matcher):
                self.report_change(original_node)
                return self.combine_calls(updated_node.left, updated_node.right)

            if self.matches_call_or_boolop(updated_node, call_matcher):
                self.report_change(original_node)
                return self.combine_call_or_boolop_fold_right(updated_node)

            if self.matches_boolop_or_call(updated_node, call_matcher):
                self.report_change(original_node)
                return self.combine_boolop_or_call_fold_left(updated_node)

        return updated_node

    def make_call_matcher(self, func_name: str) -> m.Call:
        raise NotImplementedError("Subclasses must implement this method")

    def check_calls_same_instance(
        self, left_call: cst.Call, right_call: cst.Call
    ) -> bool:
        raise NotImplementedError("Subclasses must implement this method")

    def matches_call_or_call(
        self, node: cst.BooleanOperation, call_matcher: m.Call
    ) -> bool:
        # Check for simple case: x.startswith("...") or x.startswith("...")
        call_or_call = m.BooleanOperation(
            left=call_matcher, operator=m.Or(), right=call_matcher
        )
        # True if the node matches the pattern and the calls are the same instance
        return m.matches(node, call_or_call) and self.check_calls_same_instance(
            node.left, node.right
        )

    def matches_call_or_boolop(
        self, node: cst.BooleanOperation, call_matcher: m.Call
    ) -> bool:

        # Check for case when call on left and call on left of right boolop can be combined like:
        # x.startswith("...") or x.startswith("...") and/or <any>
        call_or_boolop = m.BooleanOperation(
            left=call_matcher,
            operator=m.Or(),
            right=m.BooleanOperation(left=call_matcher),
        )
        # True if the node matches the pattern and the calls are the same instance
        return m.matches(node, call_or_boolop) and self.check_calls_same_instance(
            node.left, node.right.left
        )

    def matches_boolop_or_call(
        self, node: cst.BooleanOperation, call_matcher: m.Call
    ) -> bool:

        # Check for case when call on right and call on right of left boolop can be combined like:
        # <any> and/or x.startswith("...") or x.startswith("...")
        boolop_or_call = m.BooleanOperation(
            left=m.BooleanOperation(right=call_matcher),
            operator=m.Or(),
            right=call_matcher,
        )
        # True if the node matches the pattern and the calls are the same instance
        return m.matches(node, boolop_or_call) and self.check_calls_same_instance(
            node.left.right, node.right
        )

    def combine_calls(self, *calls: cst.Call) -> cst.Call:
        first_call = calls[0]
        new_args = []
        for arg_index in sorted(self.args_to_keep_as_is + self.args_to_combine):
            if arg_index in self.args_to_combine:
                new_args.append(self.combine_args(*calls, arg_index=arg_index))
            else:
                new_args.append(first_call.args[arg_index])

        return cst.Call(func=first_call.func, args=new_args)

    def combine_args(self, *calls: cst.Call, arg_index: int) -> cst.Arg:
        elements = []
        seen_values = set()
        for call in calls:
            arg_value = call.args[arg_index].value
            arg_elements = (
                arg_value.elements
                if isinstance(arg_value, cst.Tuple)
                else (cst.Element(value=arg_value),)
            )

            for element in arg_elements:
                if (
                    value := getattr(element.value, self.dedupilcation_attr, None)
                ) in seen_values:
                    # If an element has a non-None value that has already been seen, continue to avoid duplicates
                    continue
                if value is not None:
                    seen_values.add(value)
                elements.append(element)

        return cst.Arg(value=cst.Tuple(elements=elements))

    def combine_call_or_boolop_fold_right(
        self, node: cst.BooleanOperation
    ) -> cst.BooleanOperation:
        new_left = self.combine_calls(node.left, node.right.left)
        new_right = node.right.right
        return cst.BooleanOperation(
            left=new_left, operator=node.right.operator, right=new_right
        )

    def combine_boolop_or_call_fold_left(
        self, node: cst.BooleanOperation
    ) -> cst.BooleanOperation:
        new_left = node.left.left
        new_right = self.combine_calls(node.left.right, node.right)
        return cst.BooleanOperation(
            left=new_left, operator=node.left.operator, right=new_right
        )

clavedeluna commented on src/core_codemods/combine_calls_base.py:
```diff
@@ -0,0 +1,139 @@
+    def matches_call_or_call(
+        self, node: cst.BooleanOperation, call_matcher: m.Call
+    ) -> bool:
+        # Check for simple case: x.startswith("...") or x.startswith("...")
```
all comments in this module related to startswith/endswith can be removed

LucasFaudman replied:
Sounds good, done.

clavedeluna reviewed (approved):
approving but I know @drdavella will want to 👁️ for documentation

LucasFaudman committed:
Merge branch 'pixee:main' into combine-isinstance-issubclass

Commit:
remove startswith/endswith comments per @clavedeluna request
